runelite-client/src/main/java/net/runelite/client/plugins/puzzlesolver/solver/pathfinding/IDAStarMM.java

```diff
@@ -0,0 +1,838 @@
+
+		//Check if already placed correct
+		if (locVal.equals(locTarget))
+		{
+			return;
+		}
+
+		//Check if simple swap is enough
+		if (locBlank.equals(locTarget) && locVal.getX() - 1 == locTarget.getX())
+		{
+			swap(locBlank, locVal);
+			return;
+		}
+
+		//Move left
+		while (true)
+		{
+			locVal = findPiece(valTarget);
+			locBlank = findPiece(BLANK_TILE_VALUE);
+
+			//Check if already placed correct
+			if (locVal.equals(locTarget))
+			{
+				return;
+			}
+
+			if (locVal.getX() == locBlank.getX())
+			{
+				int diff = locBlank.getY() - locVal.getY();
+				if (diff == 1)
+				{
+					//Below
+					Point loc1 = new Point(locBlank.getX() - 1, locBlank.getY());
+					Point loc2 = new Point(loc1.getX(), loc1.getY() - 1);
+
+					swap(locBlank, loc1);
+					swap(loc1, loc2);
+					swap(loc2, locVal);
+				}
+				else if (diff == -1)
+				{
+					//Above
+
+					//Don't remove correct pices from row
+					if (locVal.getX() - 1 == locTarget.getX())
+					{
+						//Swap blank to right and continue
+						Point loc1 = new Point(locBlank.getX() + 1, locBlank.getY());
+						Point loc2 = new Point(loc1.getX(), loc1.getY() + 1);
+
+						swap(locBlank, loc1);
+						swap(loc1, loc2);
+
+						continue;
+					}
+
+					Point loc1 = new Point(locBlank.getX() - 1, locBlank.getY());
+					Point loc2 = new Point(loc1.getX(), loc1.getY() + 1);
+
+					swap(locBlank, loc1);
+					swap(loc1, loc2);
+					swap(loc2, locVal);
+				}
+			}
+			else if (locVal.getY() == locBlank.getY())
+			{
+				int diff = locBlank.getX() - locVal.getX();
+				if (diff == 1)
+				{
+					//Right
+
+					//Last piece
+					if (locTarget.getY() == DIMENSION - 1)
+					{
+						Point loc1 = new Point(locVal.getX() - 1, locVal.getY() - 1);
+						Point loc2 = new Point(locVal.getX() - 1, locVal.getY());
+						Point loc3 = new Point(locVal.getX(), locVal.getY() - 1);
+						Point loc4 = new Point(locVal.getX() + 1, locVal.getY() - 1);
+
+						swap(locBlank, locVal);
+
+						swap(locVal, loc3);
+						swap(loc3, loc1);
+						swap(loc1, loc2);
+						swap(loc2, locVal);
+						swap(locVal, loc3);
+						swap(loc3, loc1);
+						swap(loc1, loc2);
+						swap(loc2, locVal);
+						swap(locVal, locBlank);
+						swap(locBlank, loc4);
+						swap(loc4, loc3);
+						swap(loc3, loc1);
+						swap(loc1, loc2);
+						swap(loc2, locVal);
+
+						return;
+					}
+
+					Point loc1 = new Point(locVal.getX() + 1, locVal.getY() + 1);
+					Point loc2 = new Point(locVal.getX(), locVal.getY() + 1);
+					Point loc3 = new Point(locVal.getX() - 1, locVal.getY() + 1);
+					Point loc4 = new Point(locVal.getX() - 1, locVal.getY());
+
+					swap(locBlank, loc1);
+					swap(loc1, loc2);
+					swap(loc2, loc3);
+					swap(loc3, loc4);
+					swap(loc4, locVal);
+				}
+				else if (diff == -1)
+				{
+					//Left
+					swap(locBlank, locVal);
+				}
+			}
+		}
+	}
+
+	private void swap(Point p1, Point p2)
+	{
+		int[] pieces = currentState.getPuzzle().clone();
```

Comment:
Please make a method in PuzzleState that creates a copy of the state, does the swap and then returns the new PuzzleState. Then remove the getPuzzle method. This way we don't expose the puzzle array, as we want to keep it immutable.